flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobVertexBackPressureHandler.java

```diff
@@ -49,50 +54,78 @@
-                                                        ::createJobVertexBackPressureInfo)
-                                        .orElse(JobVertexBackPressureInfo.deprecated()));
+
+        TaskMetricStore taskMetricStore =
+                metricFetcher
+                        .getMetricStore()
+                        .getTaskMetricStore(jobId.toString(), jobVertexId.toString());
+
+        return CompletableFuture.completedFuture(
+                taskMetricStore != null
+                        ? createJobVertexBackPressureInfo(
+                                taskMetricStore.getAllSubtaskMetricStores())
+                        : JobVertexBackPressureInfo.deprecated());
     }
 
-    private static JobVertexBackPressureInfo createJobVertexBackPressureInfo(
-            final OperatorBackPressureStats operatorBackPressureStats) {
+    private JobVertexBackPressureInfo createJobVertexBackPressureInfo(
+            Collection<ComponentMetricStore> allSubtaskMetricStores) {
+
         return new JobVertexBackPressureInfo(
                 JobVertexBackPressureInfo.VertexBackPressureStatus.OK,
-                getBackPressureLevel(operatorBackPressureStats.getMaxBackPressureRatio()),
-                operatorBackPressureStats.getEndTimestamp(),
-                IntStream.range(0, operatorBackPressureStats.getNumberOfSubTasks())
-                        .mapToObj(
-                                subtask -> {
-                                    final double backPressureRatio =
-                                            operatorBackPressureStats.getBackPressureRatio(subtask);
-                                    return new JobVertexBackPressureInfo.SubtaskBackPressureInfo(
-                                            subtask,
-                                            getBackPressureLevel(backPressureRatio),
-                                            backPressureRatio);
-                                })
-                        .collect(Collectors.toList()));
+                getBackPressureLevel(getMaxBackPressureRatio(allSubtaskMetricStores)),
+                metricFetcher.getLastUpdateTime(),
+                createSubtaskBackPressureInfo(allSubtaskMetricStores));
+    }
+
+    private List<SubtaskBackPressureInfo> createSubtaskBackPressureInfo(
+            Collection<ComponentMetricStore> subtaskMetricStores) {
+        List<SubtaskBackPressureInfo> result = new ArrayList<>(subtaskMetricStores.size());
+        int subTaskIndex = 0;
+        for (ComponentMetricStore subtaskMetricStore : subtaskMetricStores) {
+            double backPressureRatio = getBackPressureRatio(subtaskMetricStore);
+            result.add(
+                    new SubtaskBackPressureInfo(
+                            subTaskIndex,
+                            getBackPressureLevel(backPressureRatio),
+                            backPressureRatio));
+            subTaskIndex++;
```

Comment:
this could be incorrect if at this point in time only some of the subtasks have reported back. We could change taskMetricStore.getAllSubtaskMetricStores() to also return the subtask index, or add a dedicated SubtaskMetricStore with a getter for the index.

Reply:
Good catch, I didn't know about this `MetricStore` feature/behaviour. I don't know why I was assuming it would return empty metrics in that case.

Anyway, fixed with `getAllSubtaskMetricStores` returning the a map right now. I've also extended test case to cover for that.